homeassistant/components/qnap_qsw/config_flow.py

```diff
@@ -63,3 +67,27 @@ async def async_step_user(
             ),
             errors=errors,
         )
+
+    async def async_step_dhcp(self, discovery_info: dhcp.DhcpServiceInfo) -> FlowResult:
+        """Handle DHCP discovery."""
+        ip = discovery_info.ip
+        mac = discovery_info.macaddress
+        _LOGGER.debug("DHCP discovery detected QSW: %s", mac)
+        return await self._process_discovery(mac, ip)
+
+    async def _process_discovery(self, mac: str, ip: str) -> FlowResult:
+        """Handle discovery of a gateway."""
+        url = f"http://{ip}"
+        options = ConnectionOptions(url, "", "")
+        qsw = QnapQswApi(aiohttp_client.async_get_clientsession(self.hass), options)
+
+        try:
+            await qsw.get_live()
+        except QswError as err:
+            raise AbortFlow("cannot_connect") from err
+
+        await self.async_set_unique_id(format_mac(mac))
+        self._abort_if_unique_id_configured()
+
+        self.context[CONF_URL] = url
+        return await self.async_step_user()
```

Comment:
Usually we don't send them right to the user flow if since it will ask them for data we already have (the url)

Reply:
`elkm1` has an example

Reply:
If you really want to branch in the user flow, axis/config_flow.py has an example of that, but I generally prefer to avoid branching inside a step as it creates higher risk of bugs

Reply:
Changed in d9e7d8a86e74ed5203f14a0ba4a92f28714aa6cf